tools/fp20compiler/rc1.0_combiners.cpp

```diff
@@ -53,29 +48,43 @@ void CombinersStruct::Invoke()
         //   will emit its own constants (for FACTOR#_EACH_STAGE).
         // Also see mode selection in GeneralCombinersStruct::Invoke() and
         // local-constant emitter in GeneralCombinerStruct::Invoke(int stage).
-        if (generals.localConsts == 0) {
-            printf("pb_push1(p, %s,", general_cmd);
-            printf("\n    MASK(0xFF000000, 0x%02X)", (unsigned char)(cc[i].v[3] * 0xFF));
-            printf("\n    | MASK(0x00FF0000, 0x%02X)", (unsigned char)(cc[i].v[0] * 0xFF));
-            printf("\n    | MASK(0x0000FF00, 0x%02X)", (unsigned char)(cc[i].v[1] * 0xFF));
-            printf("\n    | MASK(0x000000FF, 0x%02X)", (unsigned char)(cc[i].v[2] * 0xFF));
+        if (localConstCount == 0) {
+            printf("pb_push1(p, %s,", cmd);
+            printf("\n    MASK(0xFF000000, 0x%02X)", (unsigned char)(generalCcs.cc[i].v[3] * 0xFF));
+            printf("\n    | MASK(0x00FF0000, 0x%02X)", (unsigned char)(generalCcs.cc[i].v[0] * 0xFF));
+            printf("\n    | MASK(0x0000FF00, 0x%02X)", (unsigned char)(generalCcs.cc[i].v[1] * 0xFF));
+            printf("\n    | MASK(0x000000FF, 0x%02X)", (unsigned char)(generalCcs.cc[i].v[2] * 0xFF));
             printf(");\n");
             printf("p += 2;\n");
         }
+    }
+
+    generals.Invoke();
 
-        // Global-constants are also used in final-combiner
-        printf("pb_push1(p, %s,", final_cmd);
-        printf("\n    MASK(0xFF000000, 0x%02X)", (unsigned char)(cc[i].v[3] * 0xFF));
-        printf("\n    | MASK(0x00FF0000, 0x%02X)", (unsigned char)(cc[i].v[0] * 0xFF));
-        printf("\n    | MASK(0x0000FF00, 0x%02X)", (unsigned char)(cc[i].v[1] * 0xFF));
-        printf("\n    | MASK(0x000000FF, 0x%02X)", (unsigned char)(cc[i].v[2] * 0xFF));
+    assert(finalCcs.numConsts <= 2);
+    for (int i = 0; i < finalCcs.numConsts; i++) {
+        const char* cmd = NULL;
+        switch(finalCcs.cc[i].reg.bits.name) {
+        case REG_CONSTANT_COLOR0:
+            cmd = "NV097_SET_SPECULAR_FOG_FACTOR + 0";
+            break;
+        case REG_CONSTANT_COLOR1:
+            cmd = "NV097_SET_SPECULAR_FOG_FACTOR + 4";
+            break;
+        default:
+            assert(false);
+            break;
+        }
+
+        printf("pb_push1(p, %s,", cmd);
+        printf("\n    MASK(0xFF000000, 0x%02X)", (unsigned char)(finalCcs.cc[i].v[3] * 0xFF));
+        printf("\n    | MASK(0x00FF0000, 0x%02X)", (unsigned char)(finalCcs.cc[i].v[0] * 0xFF));
+        printf("\n    | MASK(0x0000FF00, 0x%02X)", (unsigned char)(finalCcs.cc[i].v[1] * 0xFF));
+        printf("\n    | MASK(0x000000FF, 0x%02X)", (unsigned char)(finalCcs.cc[i].v[2] * 0xFF));

```

Comment:
Create some `generalCcs.Invoke(i, REG_CONSTANT_COLOR0)` (general-combiner) and `finalCcs.Invoke(REG_CONSTANT_COLOR0)` (final-combiner)